File: actuar/R/exp.R
### ===== actuar: an R package for Actuarial Science =====
###
### Definition of the {m,lev}exp functions. The Exponential
### distribution used in these functions has cumulative distribution
### function
###
###   Pr[X <= x] = 1 - exp(-x / scale), x > 0.
###
### See Appendix A of Klugman, Panjer & Willmot, Loss Models, Second
### Edition, Wiley, 2004.
###
### AUTHORS:  Mathieu Pigeon, Vincent Goulet <vincent.goulet@act.ulaval.ca>

mexp <- function(k, rate, log = FALSE)
    .External("do_dpq", "mexp", k, rate, log)

levexp <- function(d, rate, order = 1, log = FALSE)
    .External("do_dpq", "levexp", d, rate, order, log)

File: actuar/src/exp.c
/*  ===== actuar: an R package for Actuarial Science =====
 *
 *  Functions to calculate raw and limited moments for the Exponential
 *  distribution.
 *
 *  AUTHORS: Mathieu Pigeon and Vincent Goulet <vincent.goulet@act.ulaval.ca>
 */

#include <R.h>
#include <Rmath.h>
#include "locale.h"
#include "dpq.h"

double mexp(double order, double scale, int give_log)
{
    if (!R_FINITE(scale) ||
	!R_FINITE(order) ||
	scale <= 0.0 ||
	order <= -1.0)
	return R_NaN;

    return R_pow(scale, order) * gammafn(1.0 + order);
}

double levexp(double limit, double scale, double order, int give_log)
{
    double tmp;

    if (!R_FINITE(scale) ||
	!R_FINITE(limit) ||
	!R_FINITE(order) ||
	scale <= 0.0 ||
	limit <= 0.0 ||
	order <= -1.0)
	return R_NaN;

    tmp = 1.0 + order;

    return R_pow(scale, order) * gammafn(tmp) *
	pgamma(limit, tmp, scale, 1, 0) +
	R_VG__0(limit, order) * exp(-limit * scale);
}
